src/ingest-pipeline/airflow/dags/validate_upload.py

```diff
@@ -15,166 +15,188 @@
+        print(f"lz path: {lz_path}")
+        kwargs["ti"].xcom_push(key="lz_path", value=lz_path)
+        kwargs["ti"].xcom_push(key="uuid", value=uuid)
 
     t_find_uuid = PythonOperator(
-        task_id='find_uuid',
+        task_id="find_uuid",
         python_callable=find_uuid,
         provide_context=True,
         op_kwargs={
-            }
-        )
+            "crypt_auth_tok": (
+                encrypt_tok(airflow_conf.as_dict()["connections"]["APP_CLIENT_SECRET"]).decode()
+            )
+        },
+    )
 
     def run_validation(**kwargs):
-        lz_path = kwargs['ti'].xcom_pull(key='lz_path')
-        uuid = kwargs['ti'].xcom_pull(key='uuid')
+        # Force an exception to test internal email
+        # raise Exception("test")
+        # Force an exception to test internal and external email
+        # raise ValueError
+        lz_path = kwargs["ti"].xcom_pull(key="lz_path")
+        uuid = kwargs["ti"].xcom_pull(key="uuid")
         plugin_path = [path for path in ingest_validation_tests.__path__][0]
 
-        ignore_globs = [uuid, 'extras', '*metadata.tsv',
-                        'validation_report.txt']
+        ignore_globs = [uuid, "extras", "*metadata.tsv", "validation_report.txt"]
         #
         # Uncomment offline=True below to avoid validating orcid_id URLs &etc
         #
         upload = ingest_validation_tools_upload.Upload(
             directory_path=Path(lz_path),
             dataset_ignore_globs=ignore_globs,
-            upload_ignore_globs='*',
+            upload_ignore_globs="*",
             plugin_directory=plugin_path,
             # offline=True,  # noqa E265
             add_notes=False,
-            extra_parameters={'coreuse': get_threads_resource('validate_upload', 'run_validation')}
+            extra_parameters={
+                "coreuse": get_threads_resource("validate_upload", "run_validation")
+            },
         )
         # Scan reports an error result
         report = ingest_validation_tools_error_report.ErrorReport(
-            errors=upload.get_errors(plugin_kwargs=kwargs),
-            info=upload.get_info()
+            errors=upload.get_errors(plugin_kwargs=kwargs), info=upload.get_info()
         )
-        validation_file_path = Path(get_tmp_dir_path(kwargs['run_id'])) / 'validation_report.txt'
-        with open(validation_file_path, 'w') as f:
+        validation_file_path = Path(get_tmp_dir_path(kwargs["run_id"])) / "validation_report.txt"
+        with open(validation_file_path, "w") as f:
             f.write(report.as_text())
-        kwargs['ti'].xcom_push(key='validation_file_path', value=str(validation_file_path))
+        kwargs["ti"].xcom_push(key="validation_file_path", value=str(validation_file_path))
 
     t_run_validation = PythonOperator(
-        task_id='run_validation',
+        task_id="run_validation",
         python_callable=run_validation,
         provide_context=True,
         op_kwargs={
-        }
+            "crypt_auth_tok": (
```

Comment:
Is this really necessary? the token is already there in the DAG, you can use it as it is used in send_status_msg

Reply:
I struggled a little bit with getting the crypt_auth_tok to the failure callback. The failure callback receives context as its only argument, which does not include the dag's kwargs. I could xcom_push crypt_auth_tok but that would mean that, if the dag failed prior to that point, it would not have the token. I could do an xcom_push in the first line of the first function (find_uuid) instead of requiring op_kwargs to be passed to every task; this would have the benefit of being a single line fix, and should limit the cases where the dag can fail without first passing the necessary token to the failure callback. Of course, I could be missing something here! Please let me know what you think.